c/tskit/trees.c

```diff
@@ -4551,3 +4411,224 @@ tsk_diff_iter_next(tsk_diff_iter_t *self, double *ret_left, double *ret_right,
     self->tree_left = right;
     return ret;
 }
+
+typedef struct {
+    int *m;
+    double *M;
+    tsk_id_t n;
+    tsk_id_t N;
+} kc_vectors;
+
+static int
+kc_vectors_alloc(kc_vectors *self, tsk_id_t n)
+{
+    int ret = 0;
+
+    self->n = n;
+    self->N = (n * (n - 1)) / 2;
+    self->m = calloc((size_t)(self->N + self->n), sizeof(int));
+    self->M = calloc((size_t)(self->N + self->n), sizeof(double));
+    if (self->m == NULL || self->M == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+out:
+    return ret;
+}
+
+static void
+kc_vectors_free(kc_vectors *self)
+{
+    tsk_safe_free(self->m);
+    tsk_safe_free(self->M);
+}
+
+static void
+update_kc_vectors_single_leaf(
+    tsk_tree_t *tree, kc_vectors *kc_vecs, tsk_id_t u, double time)
+{
+    const tsk_id_t *sample_index_map = tree->tree_sequence->sample_index_map;
```

Comment:
This should take a tree sequence instead of a tree because it doesn’t use the tree and can then be reused in incremental algorithms where we have the sequence but no actual tree (same with similar helpers)

Reply:
Since it’s a static method on a tree sequence, do we give it the same name spacing tsk_update...?

Reply:
Sure, let's pass in the tree sequence. I don't think we need to make it a method of tsk_treeseq, since it doesn't alter any state. The name is fine as it's static and won't be exported.

Reply:
Might be worth inlining these helper functions too, as they are so simple.

Reply:
sounds good!